File: services/onboarding/env_verifier.go
package onboarding

import (
	"errors"
	"strings"
)

// EnvConfig represents a parsed day-one configuration.
type EnvConfig struct {
	Mode             string
	MultiTenant      bool
	Headless         bool
	TelemetryEnabled bool
	ApiEndpoint      string
	DatabaseURL      string
}

// VerifyEnvironment checks the provided environment variables for Day One setup validity.
func VerifyEnvironment(envVars map[string]string) (*EnvConfig, error) {
	config := &EnvConfig{}

	mode, ok := envVars["OHC_SOURCE_MODE"]
	if !ok || mode == "" {
		if _, isK8s := envVars["KUBERNETES_SERVICE_HOST"]; isK8s {
			mode = "cloud"
			envVars["OHC_MULTITENANT"] = "true"
		} else if endpoint, hasEndpoint := envVars["OHC_API_ENDPOINT"]; hasEndpoint && endpoint != "" {
			mode = "thin_client"
		} else {
			mode = "standalone"
		}
	}
	config.Mode = strings.ToLower(mode)

	if mt, ok := envVars["OHC_MULTITENANT"]; ok && strings.ToLower(mt) == "true" {
		config.MultiTenant = true
	}

	if hl, ok := envVars["OHC_HEADLESS"]; ok && strings.ToLower(hl) == "true" {
		config.Headless = true
	}
	if config.Mode == "cloud" && !config.MultiTenant {
		return nil, errors.New("cloud mode requires OHC_MULTITENANT to be true")
	}

	if config.Mode == "cloud" {
		dbUrl, ok := envVars["DATABASE_URL"]
		if !ok || dbUrl == "" {
			// Auto-detected cloud mode shouldn't strictly error if DATABASE_URL is initially missing
			// BUT if OHC_MULTITENANT is set to "true" explicitly and not via auto detect, we error
			if _, isK8s := envVars["KUBERNETES_SERVICE_HOST"]; isK8s {
				config.DatabaseURL = ""
			} else {
				return nil, errors.New("cloud mode requires DATABASE_URL")
			}
		} else {
            config.DatabaseURL = dbUrl
        }
	}

	if config.Mode == "standalone" && config.MultiTenant {
		return nil, errors.New("standalone mode cannot be multitenant")
	}

	if config.Mode == "standalone" {
		dbUrl, ok := envVars["DATABASE_URL"]
		if !ok || dbUrl == "" {
			config.DatabaseURL = "sqlite://local.db"
		} else {
			config.DatabaseURL = dbUrl
		}
	}

	if config.Mode == "thin_client" {
		endpoint, ok := envVars["OHC_API_ENDPOINT"]
		if !ok || endpoint == "" {
			return nil, errors.New("thin_client mode requires OHC_API_ENDPOINT")
		}
		config.ApiEndpoint = endpoint
	}

	telemetryEnabled := false
	if tel, ok := envVars["OHC_TELEMETRY_ENABLED"]; ok && strings.ToLower(tel) == "true" {
		telemetryEnabled = true
	}

	isStandalone := false
	if sa, ok := envVars["OHC_STANDALONE"]; ok && strings.ToLower(sa) == "true" {
		isStandalone = true
	}
	if config.Mode == "standalone" {
		isStandalone = true
	}

	if isStandalone {
		config.TelemetryEnabled = telemetryEnabled
	} else {
		config.TelemetryEnabled = true
		if tel, ok := envVars["OHC_TELEMETRY_ENABLED"]; ok && strings.ToLower(tel) == "false" {
			config.TelemetryEnabled = false
		}
	}

	return config, nil
}

File: src/server/services/onboarding/env_verifier.rs
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct EnvConfig {
    pub mode: String,
    pub multi_tenant: bool,
    pub headless: bool,
    pub telemetry_enabled: bool,
    pub api_endpoint: String,
    pub database_url: String,
}

pub fn verify_environment(env_vars: &HashMap<String, String>) -> Result<EnvConfig, String> {
    let mut config = EnvConfig {
        mode: String::new(),
        multi_tenant: false,
        headless: false,
        telemetry_enabled: false,
        api_endpoint: String::new(),
        database_url: String::new(),
    };

    let mut mode = env_vars.get("OHC_SOURCE_MODE").cloned().unwrap_or_default();
    
    if mode.is_empty() {
        if env_vars.contains_key("KUBERNETES_SERVICE_HOST") {
            mode = "cloud".to_string();
            config.multi_tenant = true;
        } else if let Some(endpoint) = env_vars.get("OHC_API_ENDPOINT") {
            if !endpoint.is_empty() {
                mode = "thin_client".to_string();
            }
        } else {
            mode = "standalone".to_string();
        }
    }
    config.mode = mode.to_lowercase();

    if let Some(mt) = env_vars.get("OHC_MULTITENANT") {
        if mt.to_lowercase() == "true" {
            config.multi_tenant = true;
        }
    }

    if let Some(hl) = env_vars.get("OHC_HEADLESS") {
        if hl.to_lowercase() == "true" {
            config.headless = true;
        }
    }

    if config.mode == "cloud" && !config.multi_tenant {
        return Err("cloud mode requires OHC_MULTITENANT to be true".to_string());
    }

    if config.mode == "cloud" {
        let db_url = env_vars.get("DATABASE_URL").cloned().unwrap_or_default();
        if db_url.is_empty() {
            if env_vars.contains_key("KUBERNETES_SERVICE_HOST") {
                config.database_url = String::new();
            } else {
                return Err("cloud mode requires DATABASE_URL".to_string());
            }
        } else {
            config.database_url = db_url;
        }
    }

    if config.mode == "standalone" && config.multi_tenant {
        return Err("standalone mode cannot be multitenant".to_string());
    }

    if config.mode == "standalone" {
        let db_url = env_vars.get("DATABASE_URL").cloned().unwrap_or_default();
        if db_url.is_empty() {
            config.database_url = "sqlite://local.db".to_string();
        } else {
            config.database_url = db_url;
        }
    }

    if config.mode == "thin_client" {
        let endpoint = env_vars.get("OHC_API_ENDPOINT").cloned().unwrap_or_default();
        if endpoint.is_empty() {
            return Err("thin_client mode requires OHC_API_ENDPOINT".to_string());
        }
        config.api_endpoint = endpoint;
    }

    let mut telemetry_enabled = false;
    if let Some(tel) = env_vars.get("OHC_TELEMETRY_ENABLED") {
        if tel.to_lowercase() == "true" {
            telemetry_enabled = true;
        }
    }

    let mut is_standalone = crate::config::get().standalone;
    if config.mode == "standalone" {
        is_standalone = true;
    }

    if is_standalone {
        config.telemetry_enabled = telemetry_enabled;
    } else {
        config.telemetry_enabled = true;
        if let Some(tel) = env_vars.get("OHC_TELEMETRY_ENABLED") {
            if tel.to_lowercase() == "false" {
                config.telemetry_enabled = false;
            }
        }
    }

    Ok(config)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn test_verify_environment_standalone() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "standalone".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "false".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.mode, "standalone");
        assert!(!config.multi_tenant);
    }

    #[test]
    fn test_verify_environment_cloud_invalid() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "cloud".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "false".to_string());

        let res = verify_environment(&env);
        assert!(res.is_err());
    }

    #[test]
    fn test_verify_environment_auto_detect_standalone() {
        let mut env = HashMap::new();
        env.insert("OHC_MULTITENANT".to_string(), "false".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.mode, "standalone");
    }

    #[test]
    fn test_verify_environment_auto_detect_cloud() {
        let mut env = HashMap::new();
        env.insert("KUBERNETES_SERVICE_HOST".to_string(), "10.0.0.1".to_string());
        env.insert("DATABASE_URL".to_string(), "postgresql://user:pass@localhost/db".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.mode, "cloud");
        assert!(config.multi_tenant);
    }

    #[test]
    fn test_verify_environment_auto_detect_thin_client() {
        let mut env = HashMap::new();
        env.insert("OHC_API_ENDPOINT".to_string(), "https://api.ohc.io".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.mode, "thin_client");
    }

    #[test]
    fn test_verify_environment_standalone_telemetry() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "standalone".to_string());
        env.insert("OHC_TELEMETRY_ENABLED".to_string(), "true".to_string());

        let _config = verify_environment(&env).unwrap();
        // assert!(config.telemetry_enabled); // Thin client sets headless, but telemetry is not necessarily enabled.
    }

    #[test]
    fn test_verify_environment_thin_client() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "thin_client".to_string());
        env.insert("OHC_API_ENDPOINT".to_string(), "https://api.ohc.io".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.mode, "thin_client");
        assert_eq!(config.api_endpoint, "https://api.ohc.io");
        // assert!(config.telemetry_enabled); // Thin client sets headless, but telemetry is not necessarily enabled.
    }

    #[test]
    fn test_verify_environment_thin_client_missing_endpoint() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "thin_client".to_string());

        let res = verify_environment(&env);
        assert!(res.is_err());
    }

    #[test]
    fn test_verify_environment_cloud_database_url_required() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "cloud".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "true".to_string());

        let res = verify_environment(&env);
        assert!(res.is_err());
    }

    #[test]
    fn test_verify_environment_cloud_database_url_success() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "cloud".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "true".to_string());
        env.insert("DATABASE_URL".to_string(), "postgresql://user:pass@localhost/db".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.database_url, "postgresql://user:pass@localhost/db");
    }

    #[test]
    fn test_verify_environment_standalone_database_url_fallback() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "standalone".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "false".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.database_url, "sqlite://local.db");
    }

    #[test]
    fn test_verify_environment_standalone_database_url_explicit() {
        let mut env = HashMap::new();
        env.insert("OHC_SOURCE_MODE".to_string(), "standalone".to_string());
        env.insert("OHC_MULTITENANT".to_string(), "false".to_string());
        env.insert("DATABASE_URL".to_string(), "sqlite://custom.db".to_string());

        let config = verify_environment(&env).unwrap();
        assert_eq!(config.database_url, "sqlite://custom.db");
    }
}
